File: agama/make_nfw.py
#!/usr/bin/env python

import agama
import argparse
from math import pi

from h5py import File


def main():
    args = parse_args()
    pos, vel, mass = sample_nfw(N=args.N, verbose=args.verbose, cutoff=args.truncation_radius)
    write_to_fits(pos, vel, mass, args.output)

def parse_args():
    parser = argparse.ArgumentParser(description='Create NFW potential')
    parser.add_argument('-n', '--N', type=float, default=1e4, help='Number of particles')
    parser.add_argument('-o', '--output', type=str, default='nfw.hdf5', help='Output file name')
    parser.add_argument('-v', '--verbose', action='store_true', help='Verbose output')
    parser.add_argument('-t', '--truncation_radius', type=float, default=64, help='Truncation radius over scale radius')

    return parser.parse_args()



def sample_nfw(N=1e4, verbose=False, cutoff=None):
    # Create a NFW potential
    if not isinstance(N, (int, float)):
        raise ValueError('N must be a number')


    if N % 1 != 0:
        raise ValueError('N must be an integer')

    if N < 1:
        raise ValueError('N must be greater than 1')

    N = int(N)
    scaleRadius = 1
    mass = 1

    rho0 = mass / (4*pi * scaleRadius**3 ) / 3

    pot = agama.Potential(type='Spheroid', densityNorm=rho0,
        gamma=1, beta=3, alpha=1, scaleRadius=scaleRadius,
        outerCutoffRadius = cutoff*scaleRadius,
        cutoffStrength=1)

    df = agama.DistributionFunction(type="QuasiSpherical", potential=pot)
    gm = agama.GalaxyModel(pot, df)

    posvel, mass = gm.sample(N)
    pos = posvel[:,:3]
    vel = posvel[:,3:]

    return pos, vel, mass



def write_to_hdf5(pos, vel, mass, output='nfw.fits'):
    with File(output, 'w') as f:
        f.create_dataset('pos', data=pos)
        f.create_dataset('vel', data=vel)
        f.create_dataset('mass', data=mass)
    print(f'Wrote {len(pos)} particles to {output}')



if __name__ == '__main__':
    main()

File: agama/make_nfw.jl
#!/usr/bin/env julia
using PythonCall

using ArgParse
using LilGuys

agama = pyimport("agama")


function get_args()
    s = ArgParseSettings(
        description="Generate a snapshot of N particles from an NFW profile"
        )

    @add_arg_table s begin
        "output" 
            help="output file. If not provided, will be halos/N_nfw.hdf5 where N is the number of particles"
            default = nothing
        "-N", "--number"
            help="number of particles"
            default="1e4"
        "-t", "--cutoff"
            help="exponential cutoff radius in units of scale radius"
            arg_type=Float64
            default=100
        "-b", "--beta"
            help="Velocity anisotropy parameter"
            arg_type=Float64
            default=0.0
        "-a", "--r-a"
            help="Anisotropy scale radius"
            arg_type=Float64
            default=Inf
    end

    args = parse_args(s)


    if args["output"] === nothing
        filename = "halos/nfw_$(args["number"])"

        if args["beta"] != 0.0
            filename = filename * "_beta$(args["beta"])"
        end
        if args["r-a"] != Inf
            filename = filename * "_ra$(args["r-a"])"
        end

        @info "Saving to $filename.hdf5"
        args["output"] = filename * ".hdf5"
    end

    args["number"] = convert(Int, parse(Float64, args["number"]))

    if args["number"] < 1
        throw(ArgumentError("N must be positive"))
    end

    return args
end


function main()
    args = get_args()

    if isfile(args["output"])
        throw(ArgumentError("Output file already exists"))
    end

    cutoff = args["cutoff"]
    N = args["number"]

    scaleRadius = 1
    mass = 1
    rho0 = mass / (4π * scaleRadius^3) 


    pot = agama.Potential(type="Spheroid", densityNorm=rho0,
        gamma=1, beta=3, alpha=1, scaleRadius=scaleRadius,
        outerCutoffRadius = cutoff*scaleRadius,
        cutoffStrength=1)

    df = agama.DistributionFunction(type="QuasiSpherical", potential=pot,
        beta0=args["beta"], r_a=args["r-a"])

    gm = agama.GalaxyModel(pot, df)

    posvel, mass = gm.sample(N)
    posvel = pyconvert(Matrix{Float64}, posvel)
    posvel = posvel'

    pos = posvel[1:3, :]
    vel = posvel[4:6, :]
    mass = pyconvert(Float64, mass[1])


    snap = Snapshot(pos, vel, mass) 

    LilGuys.save(args["output"], snap)
end


if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
